### ISSUE TITLE
Manual Optimization does not synchronize gradients in DDP

### ISSUE THREAD

**Issue by mhamilton723** [PyTorchLightning/pytorch-lightning#9237]
## 🐛 Bug

Gradients are not synchronized properly with manual optimization in DDP

### To Reproduce

```
import pytorch_lightning as pl
import torch
from pytorch_lightning import LightningModule
from pytorch_lightning.utilities.seed import seed_everything
from torch.utils.data import DataLoader, Dataset


class RandomDataset(Dataset):
    def __init__(self, size, num_samples):
        self.len = num_samples
        self.x = torch.randn(num_samples, size)
        self.y = self.x[:, 0] > .5

    def __getitem__(self, index):
        return self.x[index], self.y[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 1)
        self.automatic_optimization = False

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        opt = self.optimizers()
        opt.zero_grad()
        print("{} {} {}".format(self.global_step, self.global_rank, self.layer.weight.reshape(-1)[0:5]))
        x, y = batch
        output = self.layer(x)
        loss = (output - y.to(torch.float32)).square().mean()

        loss.backward()
        opt.step()
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        return optimizer


seed_everything(seed=0)

# init model
model = BoringModel()

# Initialize a trainer
trainer = pl.Trainer(
    max_steps=2,
    progress_bar_refresh_rate=0,
    gpus=[0,1],
    accelerator='ddp',
)

train = RandomDataset(32, 1000)
train = DataLoader(train, batch_size=32)

# Train the model ⚡
trainer.fit(model, train)
```
Outputs:

```
0 0 tensor([-0.0013,  0.0948, -0.1455, -0.1301, -0.0681], device='cuda:0',
       grad_fn=<SliceBackward>)
0 1 tensor([-0.0013,  0.0948, -0.1455, -0.1301, -0.0681], device='cuda:1',
       grad_fn=<SliceBackward>)
1 1 tensor([-0.0155,  0.0599, -0.1349, -0.0937, -0.0467], device='cuda:1',
       grad_fn=<SliceBackward>)
1 0 tensor([-0.0187,  0.0803, -0.1109, -0.0780, -0.0527], device='cuda:0',
       grad_fn=<SliceBackward>)
```

### Expected behavior

The parameters should stay the same after a step. And do stay the same if automatic optimization is true.

### Environment

* CUDA:
        - GPU:
                - Tesla V100-PCIE-16GB
                - Tesla V100-PCIE-16GB
                - Tesla V100-PCIE-16GB
                - Tesla V100-PCIE-16GB
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.20.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.1
        - pytorch-lightning: 1.4.4
        - tqdm:              4.59.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #58~18.04.1-Ubuntu SMP Wed Jul 28 23:14:18 UTC 2021

**Comment by mhamilton723** [comment]
Update: 

switching `loss.backward()` to `self.manual_backward(loss)` has done the trick!
